Python/bytecodes.c

```diff
@@ -3015,10 +3015,20 @@ dummy_func(
 
 
         op(_ITER_CHECK_LIST, (iter -- iter)) {
-            EXIT_IF(Py_TYPE(PyStackRef_AsPyObjectBorrow(iter)) != &PyListIter_Type);
+            PyObject *iter_o = PyStackRef_AsPyObjectBorrow(iter);
+            EXIT_IF(Py_TYPE(iter_o) != &PyListIter_Type);
+#ifdef Py_GIL_DISABLED
+            _PyListIterObject *it = (_PyListIterObject *)iter_o;
+            EXIT_IF(it->it_seq == NULL ||
+                    !_Py_IsOwnedByCurrentThread((PyObject *)it->it_seq) ||
+                    !_PyObject_GC_IS_SHARED(it->it_seq));
```

Comment:
Ideally, we'd check these once on entry to the loop, not on every iteration, but that might require broader changes.

Reply:
Yeah, I intentionally left that as a potential future avenue for improvement. It's not how the specializing interpreter is set up, and we have to be pretty damned sure the iterators can't legitimately escape the loop. (It's actually pretty easy to make the interpreter do this with generators, as is, by using a weakref, and for list/tuple/range iterators I'm not certain it can't be done with instrumentation or stack inspection... So I'm not comfortable making those assumptions right now.)

Reply:
What I meant to say is that currently the specializing interpreter will not unspecialize the code when the guards fail, it will just unspecialize the current iteration (by jumping to the unspecialized code), and only checking on loop entry means we'd have to unspecialize the code.